### ISSUE TITLE
Modeling relation that doesn't follow naming strategy

### ISSUE THREAD

**Issue by emrosenf** [mikro-orm/mikro-orm#308]
My application follow the `EntityCaseNamingStrategy` and I don't wish to change that.

There are two tables that I didn't design and I would like to accommodate.

Basically `Account` has many `AccountSegment`s. Here is the abridged DDL

```sql
CREATE TABLE "Account" (
    "accountNumber" text PRIMARY KEY,
    ...
);

CREATE TABLE "AccountSegment" (
    "segmentId" text PRIMARY KEY,
    "accountNumber" text,
    ...
);
```

What options do I have to pass to `@ManyToOne` and `@OneToMany` to accommodate this? I've been trying with no luck along the lines of:

```javascript

  @ManyToOne(() => Account, { inversedBy: 'segments', })
  account?: IdentifiedReference<Account>;

  ...

  @OneToMany(() => AccountSegment, segment => segment.account, { referenceColumnName: 'accountNumber', })
  segments = new Collection<AccountSegment>(this);

**Comment by B4nan** [comment]
Settings `fieldName` on the owning side (m:1) should work. Will try tomorrow.

**Comment by B4nan** [comment]
Composite PKs are not fully supported (you can use them only for M:N pivot tables currently). I am planning to add full support for them in one of 3.x releases. 

BTW so you have composite PK where some fields of it are optional? Is that really valid (from the SQL point of view)?

**Comment by emrosenf** [comment]
Gotcha. You are right, the `nullable` was a mistake. SQL does not support NULL values in composite PKs.

Since I only need to read these objects and don't need any write abilities, for the moment I have amended [EntityValidator#66](https://github.com/mikro-orm/mikro-orm/blob/master/lib/entity/EntityValidator.ts#L66) to disable the error if there is a composite PK.

```typescript
  validatePrimaryKey<T extends AnyEntity<T>>(entity: EntityData<T>, meta: EntityMetadata): void {
    // Do not throw error if there is a composite PK
    if (!entity || (!entity[meta.primaryKey] && !entity[meta.serializedPrimaryKey] && !meta.compositePK)) {
      throw ValidationError.fromMergeWithoutPK(meta);
    }
  }
```

**Comment by B4nan** [comment]
And this throws why exactly? Because this validation will be present for composite keys as well - you cannot merge entity that uses composite PK and does not have all of them set already. The way it will be probably implemented is by introducing virtual string PK that will combine all of the composite PKs into single unique string (which will be then used as a key in the identity map). 

`meta.primaryKey` should point to one of your primary keys (probably the last defined one), right?

**Comment by emrosenf** [comment]
That's not actually the source of the error.

![Screen Shot 2020-01-23 at 8 05 43 AM](https://user-images.githubusercontent.com/425523/73001357-2eb2ca00-3db7-11ea-80f8-7e9bef941d29.png)

The problem is that `meta.primaryKey` is "accountNumber" but in the `entity` (owning side of 1:m), the field has been renamed to `account`.

**Comment by B4nan** [comment]
Can you provide full code for both entities?

**Comment by emrosenf** [comment]
```typescript

import { Property, AnyEntity, PrimaryKey, Collection, OneToMany, ManyToOne, IdentifiedReference } from 'mikro-orm';

export class BaseRow implements AnyEntity<BaseRow, 'accountNumber'> {
  constructor(args?: Partial<BaseRow>) {
    Object.assign(this, args)
  }

  @PrimaryKey({ type: 'text', })
  accountNumber: string;
  
  @Property({ type: 'text', nullable: true, })
  situsHouseNumber: string;
  
  @Property({ type: 'text', nullable: true, })
  situsStreet: string;
 
}

export class BaseSegment implements AnyEntity<BaseSegment, 'accountNumber'> {
  constructor(args?: Partial<BaseSegment>) {
    Object.assign(this, args)
  }

  @PrimaryKey({ type: 'text', })
  accountNumber: string;

  @PrimaryKey({ type: 'text' })
  improvementId: string;
  
  @PrimaryKey({ type: 'text' })
  detailId: string;
  
  @Property({ type: 'text', nullable: true, })
  type: string;
  
  @Property({ type: 'text', nullable: true, })
  description: string;

  @Property({ nullable: true, })
  yearBuilt: number;

  @Property({ nullable: true, })
  value: number;
}

@Entity({ collection: 'Row', })
export class Account extends BaseRow {

  @OneToMany(() => Segment, segment => segment.account)
  segments = new Collection<Segment>(this);

}

@Entity({ collection: 'Segment', })
export class AccountSegment extends BaseSegment {

  @ManyToOne(() => Account, { fieldName: 'accountNumber', })
  account?: IdentifiedReference<Account>;
}
```

**Comment by B4nan** [comment]
Hmm so I guess I finally understand, your `AccountSegment` entity has both `account` and `accountNumber` properties, but in fact you have only one (`accountNumber`) in the database. 

I am kinda surprised that it is working without changes, but all you need to do is to use `joinColumn` on the 1:m side, instead of the `fieldName` on m:1 side.

```typescript
@Entity({ collection: 'Row' })
export class Account extends BaseRow {

  @OneToMany(() => AccountSegment, segment => segment.account, { joinColumn: 'accountNumber' })
  segments = new Collection<AccountSegment>(this);

}

@Entity({ collection: 'Segment' })
export class AccountSegment extends BaseSegment {

  @ManyToOne(() => Account)
  account?: IdentifiedReference<Account>;

}
```

**Comment by emrosenf** [comment]
Thank again @B4nan. I actually needed to use both the `joinColumn` on the 1:m and `fieldName` on the m:1 side.

**Comment by B4nan** [comment]
Can we close this, or do you think we need to improve the docs somewhere? 

For composite pk support I will create new issue.

**Comment by emrosenf** [comment]
I deleted my cache and updated to 3.1.0 and now this isn't working for me again. (BTW I don't think v3.1.0 is the problem).

I am using `fieldName` on the m:1 side, nothing special on the 1:m side. The `entity` only has `account` and not `accountNumber` as in the [screenshot above](https://github.com/mikro-orm/mikro-orm/issues/308#issuecomment-577749248).

I definitely think more documentation is warranted about the options and what they do

**Comment by B4nan** [comment]
Can you please retest with v3.4? Could you ideally prepare a reproduction repo?

**Comment by emrosenf** [comment]
Still doesn't work. [Minimal reproduction](https://github.com/emrosenf/mikro-orm-308)

PS: congrats on all of the amazing new features you've added since I last checked

**Comment by B4nan** [comment]
Ok this helps a lot, thanks for a real [MVCE](https://stackoverflow.com/help/minimal-reproducible-example)! So the real issue is that you were trying to map single column (`AccountSegment.accountNumber`) to multiple properties (`AccountSegment.account` and `BaseSegment.accountNumber`). 

This is currently not supported, each column is mapped to a single property only. So what happens here is that your db column `accountNumber` is mapped to `account` defined at the `AccountSegment` level, but not to the `BaseSegment.accountNumber`, which is a PK. 

Honestly mapping a column to multiple properties feels like a bad design to me, there was already another issue mentioning this but there was a quite easy workaround as it was not about primary keys so one could use a getter. The reason why it feels bad is that the reverse mapping, you would have multiple properties that could potentially have conflicting options. But with the PK use case you are having it sounds more reasonable... 

A workaround for this example is to not use the `BaseSegment` entity that is causing the problem here. Something like this should work (not 100% sure if it works with 3.4.1, but it will work once #395 is merged).

```typescript
@Entity()
export class AccountSegment {

  constructor(args?: Partial<BaseSegment>) {
    Object.assign(this, args);
  }

  @ManyToOne(() => Account, { fieldName: 'accountNumber', wrappedReference: true, primary: true })
  account?: IdentifiedReference<Account>;

  @Property({ persist: false })
  get accountNumber(): number | undefined {
    console.log(this.account);
    return this.account?.unwrap().accountNumber;
  }

  @Property()
  yearBuilt: number;

}
```

Note that I also added `wrappedReference: true`, that is needed if you use `ReflectMetadataProvider` so the ORM will hydrate a reference rather than an entity directly.

**Comment by emrosenf** [comment]
Thank you for taking another look.

This doesn't work as is, but I will check again after #395 is merged.

**Comment by B4nan** [comment]
v3.5.0 is out. I was thinking about this again, and I believe you should define both `account` and `accountSegment` in the `BaseSegment` entity (in the way I described above).

**Comment by emrosenf** [comment]
I brought the definitions to the subclass level and it's working. Thanks for everything.
